awkward/array/table.py

```diff
@@ -53,18 +53,25 @@ def tolist(self):
                 return dict((n, self._table._try_tolist(x[self._index])) for n, x in self._table._contents.items())
 
         def __len__(self):
-            i = 0
-            while str(i) in self._table._contents:
-                i += 1
-            return i
+            if self._table.rowname == 'Row':
+                return len(self._table._contents)
+            else:
+                i = 0
+                while str(i) in self._table._contents:
+                    i += 1
+                return i
 
         def __iter__(self, checkiter=True):
             if checkiter:
                 self._table._checkiter()
-            i = 0
-            while str(i) in self._table._contents:
-                yield self._table._contents[str(i)][self._index]
-                i += 1
+            if self._table.rowname == 'Row':
```

Comment:
This shouldn't be a whitelist for names (`'Row'`), but a blacklist: the `rowname` can be used to customize the look of objects; I've presented examples in which the `rowname` is `"Event"` or `"Muon"`. The ultimate purpose of the `rowname` is to allow [nominative typing in a mostly structural typing system](http://wiki.c2.com/?NominativeAndStructuralTyping). Thus, `Point3D` and `Vector3D` might both have the same fields with the same names and types, but `Point3D` should be affine-transformed when changing coordinate system, while `Vector3D` is only linearly-transformed.

So we can't have `rowname == 'Row'`. If you're trying to exclude the tuple case, perhaps you should say `if i == 0` here?

Reply:
> If you're trying to exclude the tuple case, perhaps you should say `if i == 0` here?

I'm not sure what you mean here... Can we just do `if rowname != 'tuple'`?

Reply:
Oh, I see, if `i` is only ever used in Table.py to count columns. That seems pretty fragile, though.

Reply:
This is defining a policy, and I guess it should _either_ iterate over tuple items (for a `Row` interpreted as a tuple) _or_ iterate over dict keys (for any other `Row`).

Maybe a better policy would be to only do the iteration over `i` if the `rowname == "tuple"` and iterate over dict keys in any other case. Then "`tuple`" is the nominative type. Actually, I like this latter option better; it's less prone to misinterpretation (i.e. a `rowname != "tuple"` that happens to have a key named `"0"`).

Reply:
Sure, that makes sense. I've switched it around to that.